<a href="https://colab.research.google.com/github/devyanikal/Credit-Card-Fraud-Detection/blob/main/CreditCardFraudDetection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [52]:
#importing numpy and pandas libraries

import numpy as np
import pandas as pd 

Loading dataset as dataframe

In [53]:
credit_card_dataset=pd.read_csv('creditcard.csv') 

In [54]:
credit_card_dataset

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [55]:
#first 10 rows of dataset
credit_card_dataset.head(10)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   
5  0.260314 -0.568671  ... -0.208254 -0.559825 -0.026398 -0.371427 -0.232794   
6  0.081213  0.464960  ... -0.167716 -0.270710 -0.154104 -0.780055  0.750137   
7 -3.807864  0.615375  ...  1.943465 -1.015455  0.057504 -0.649709 -0.415267   
8  0.851084 -0.392048  ... -0.073425 -0.268092 -0.204233  1.011592  0.373205   
9  0.069539 -0.736727  ... -0.246914 -0.633753 -0.120794 -0.385050 -0.069733   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  
5  0.105915  0.253844  0.081080    3.67      0  
6 -0.257237  0.034507  0.005168    4.99      0  
7 -0.051634 -1.206921 -1.085339   40.80      0  
8 -0.384157  0.011747  0.142404   93.20      0  
9  0.094199  0.246219  0.083076    3.68      0  

[10 rows x 31 columns]

In [56]:
#last 10 rows of dataset
credit_card_dataset.tail(10)

Time         V1         V2        V3        V4        V5  \
284797  172782.0  -0.241923   0.712247  0.399806 -0.463406  0.244531   
284798  172782.0   0.219529   0.881246 -0.635891  0.960928 -0.152971   
284799  172783.0  -1.775135  -0.004235  1.189786  0.331096  1.196063   
284800  172784.0   2.039560  -0.175233 -1.196825  0.234580 -0.008713   
284801  172785.0   0.120316   0.931005 -0.546012 -0.745097  1.130314   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284797 -1.343668  0.929369 -0.206210  0.106234  ... -0.228876 -0.514376   
284798 -1.014307  0.427126  0.121340 -0.285670  ...  0.099936  0.337120   
284799  5.519980 -1.518185  2.080825  1.159498  ...  0.103302  0.654850   
284800 -0.726571  0.017050 -0.118228  0.435402  ... -0.268048 -0.717211   
284801 -0.235973  0.812722  0.115093 -0.204064  ... -0.314205 -0.808520   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284797  0.279598  0.371441 -0.559238  0.113144  0.131507  0.081265    5.49   
284798  0.251791  0.057688 -1.508368  0.144023  0.181205  0.215243   24.05   
284799 -0.348929  0.745323  0.704545 -0.127579  0.454379  0.130308   79.99   
284800  0.297930 -0.359769 -0.315610  0.201114 -0.080826 -0.075071    2.68   
284801  0.050343  0.102800 -0.435870  0.124079  0.217940  0.068803    2.69   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284797      0  
284798      0  
284799      0  
284800      0  
284801      0  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[10 rows x 31 columns]

Here, each row is a transaction happened in the time specified.

Eg. transaction 4 has happened 2 sec after the first transaction.

* V1 to V28 are the features of each transaction (hidden column name because credit card details are sensitive data).
* Time is in seconds.
* Amount is in US Dollars.
*It is a labeled dataset as class is provided.
   * class: 1 (fraud transaction)
   * class: 0 (legit transaction)




In [57]:
print("No. of Transactions = ",len(credit_card_dataset))

No. of Transactions =  284807


In [58]:
#dataset information
credit_card_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [59]:
#To check no. of missing values in the dataset
credit_card_dataset.isnull()

Time     V1     V2     V3     V4     V5     V6     V7     V8     V9  \
0       False  False  False  False  False  False  False  False  False  False   
1       False  False  False  False  False  False  False  False  False  False   
2       False  False  False  False  False  False  False  False  False  False   
3       False  False  False  False  False  False  False  False  False  False   
4       False  False  False  False  False  False  False  False  False  False   
...       ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
284802  False  False  False  False  False  False  False  False  False  False   
284803  False  False  False  False  False  False  False  False  False  False   
284804  False  False  False  False  False  False  False  False  False  False   
284805  False  False  False  False  False  False  False  False  False  False   
284806  False  False  False  False  False  False  False  False  False  False   

        ...    V21    V22    V23    V24    V25    V26    V27    V28  Amount  \
0       ...  False  False  False  False  False  False  False  False   False   
1       ...  False  False  False  False  False  False  False  False   False   
2       ...  False  False  False  False  False  False  False  False   False   
3       ...  False  False  False  False  False  False  False  False   False   
4       ...  False  False  False  False  False  False  False  False   False   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...     ...   
284802  ...  False  False  False  False  False  False  False  False   False   
284803  ...  False  False  False  False  False  False  False  False   False   
284804  ...  False  False  False  False  False  False  False  False   False   
284805  ...  False  False  False  False  False  False  False  False   False   
284806  ...  False  False  False  False  False  False  False  False   False   

        Class  
0       False  
1       False  
2       False  
3       False  
4       False  
...       ...  
284802  False  
284803  False  
284804  False  
284805  False  
284806  False  

[284807 rows x 31 columns]

In [60]:
credit_card_dataset.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

Note: No null values found.

In [61]:
#No. of legit and fraud transactions
credit_card_dataset['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

No. of normal transactions: 284315

No. of fraud transactions: 492

Therefore we can say that dataset is highly unbalanced.

In [62]:
legit_trans=credit_card_dataset[credit_card_dataset['Class']==0]
fraud_trans=credit_card_dataset[credit_card_dataset['Class']==1]

In [63]:
legit_trans

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284315 rows x 31 columns]

In [64]:
legit_trans.shape

(284315, 31)

In [65]:
fraud_trans

Time        V1        V2        V3        V4        V5        V6  \
541        406.0 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
623        472.0 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
4920      4462.0 -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788   
6108      6986.0 -4.397974  1.358367 -2.592844  2.679787 -1.128131 -1.706536   
6329      7519.0  1.234235  3.019740 -4.304597  4.732795  3.624201 -1.357746   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
541    -2.537387  1.391657 -2.770089  ...  0.517232 -0.035049 -0.465211   
623     0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   
4920    0.562320 -0.399147 -0.238253  ... -0.294166 -0.932391  0.172726   
6108   -3.496197 -0.248778 -0.247768  ...  0.573574  0.176968 -0.436207   
6329    1.713445 -0.496358 -1.282858  ... -0.379068 -0.704181 -0.656805   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
541     0.320198  0.044519  0.177840  0.261145 -0.143276    0.00      1  
623    -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
4920   -0.087330 -0.156114 -0.542628  0.039566 -0.153029  239.93      1  
6108   -0.053502  0.252405 -0.657488 -0.827136  0.849573   59.00      1  
6329   -1.632653  1.488901  0.566797 -0.010016  0.146793    1.00      1  
...          ...       ...       ...       ...       ...     ...    ...  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[492 rows x 31 columns]

In [66]:
fraud_trans.shape

(492, 31)

In [69]:
legit_trans.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [70]:
fraud_trans.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [73]:
#comparing mean values of both the classes
credit_card_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]



---


**Taking 492 random transactions/rows from legit transactions and joining it with 492 fraud transactions/rows so that model can learn to clearly distinguish between the two classes for new data.**

---



In [74]:
legit_trans2=legit_trans.sample(n=492)

In [76]:
legit_trans2.shape

(492, 31)

In [77]:
#joining legit_trans2 and fraud_trans dataframes
credit_card_data2=pd.concat([legit_trans2,fraud_trans],axis=0)

In [78]:
credit_card_data2.shape

(984, 31)

In [79]:
credit_card_data2.head(5)

Time        V1        V2        V3        V4        V5        V6  \
252624  155893.0 -3.096772 -2.346644 -1.318299  0.837341  1.944111 -2.559015   
160407  113346.0  1.826554  0.090643  0.243831  3.906902 -0.143234  0.908784   
155461  105488.0  2.017251 -0.536542 -1.107178 -1.526652 -0.344321 -0.927326   
261270  159939.0  1.980288 -0.414870 -0.573511  0.169346 -0.166880  0.451289   
107783   70617.0 -0.897297  0.095658  1.785049  0.405704  0.450217  0.636208   

              V7        V8        V9  ...       V21       V22       V23  \
252624  0.218493  0.052690 -0.021518  ...  0.237714  1.145760  1.733398   
160407 -0.577032  0.229840  0.083330  ... -0.267829 -0.612599  0.278443   
155461 -0.325952 -0.126134  2.858817  ... -0.524116 -1.373754  0.513815   
261270 -0.708340  0.161320  1.097338  ...  0.215152  0.849706  0.065745   
107783  0.261807  0.411953  0.137857  ...  0.126444  0.449256 -0.027538   

             V24       V25       V26       V27       V28  Amount  Class  
252624 -0.014040  0.195553 -0.085511  0.420913 -0.108632  131.35      0  
160407 -0.713179 -0.384837 -0.266499  0.026640 -0.026595   34.29      0  
155461  0.609008 -0.585113 -0.877541 -0.050999 -0.050049   30.26      0  
261270  0.242743 -0.006599 -0.236766  0.038145 -0.042855    9.99      0  
107783 -0.291228 -0.094007 -0.481131  0.156686  0.150064   61.25      0  

[5 rows x 31 columns]

In [82]:
#No. of legits and fraud transactions
credit_card_data2['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [83]:
#comparing mean values of both the classes
credit_card_data2.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92916.684959 -0.020081  0.008594  0.063925 -0.019005 -0.011373   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.005752 -0.055590  0.023107  0.010076  ... -0.003979 -0.003256   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.002173 -0.017597 -0.001738 -0.034331  0.008697 -0.001809 -0.008549   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       84.065122  
1      122.211321  

[2 rows x 30 columns]

Therefore we can say that nature of the dataset is not changed.

Separating features from the target variable



In [85]:
X=credit_card_data2.drop(columns='Class',axis=1)

In [86]:
Y=credit_card_data2['Class']

In [87]:
X

Time        V1        V2        V3        V4        V5        V6  \
252624  155893.0 -3.096772 -2.346644 -1.318299  0.837341  1.944111 -2.559015   
160407  113346.0  1.826554  0.090643  0.243831  3.906902 -0.143234  0.908784   
155461  105488.0  2.017251 -0.536542 -1.107178 -1.526652 -0.344321 -0.927326   
261270  159939.0  1.980288 -0.414870 -0.573511  0.169346 -0.166880  0.451289   
107783   70617.0 -0.897297  0.095658  1.785049  0.405704  0.450217  0.636208   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V20       V21       V22  \
252624  0.218493  0.052690 -0.021518  ... -0.082407  0.237714  1.145760   
160407 -0.577032  0.229840  0.083330  ... -0.205729 -0.267829 -0.612599   
155461 -0.325952 -0.126134  2.858817  ... -0.240388 -0.524116 -1.373754   
261270 -0.708340  0.161320  1.097338  ... -0.104965  0.215152  0.849706   
107783  0.261807  0.411953  0.137857  ... -0.009897  0.126444  0.449256   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  1.252967  0.778584 -0.319189   
280143 -1.413170  0.248525 -1.127396  ...  0.226138  0.370612  0.028234   
280149 -2.234739  1.210158 -0.652250  ...  0.247968  0.751826  0.834108   
281144 -2.208002  1.058733 -1.632333  ...  0.306271  0.583276 -0.269209   
281674  0.223050 -0.068384  0.577829  ... -0.017652 -0.164350 -0.295135   

             V23       V24       V25       V26       V27       V28  Amount  
252624  1.733398 -0.014040  0.195553 -0.085511  0.420913 -0.108632  131.35  
160407  0.278443 -0.713179 -0.384837 -0.266499  0.026640 -0.026595   34.29  
155461  0.513815  0.609008 -0.585113 -0.877541 -0.050999 -0.050049   30.26  
261270  0.065745  0.242743 -0.006599 -0.236766  0.038145 -0.042855    9.99  
107783 -0.027538 -0.291228 -0.094007 -0.481131  0.156686  0.150064   61.25  
...          ...       ...       ...       ...       ...       ...     ...  
279863  0.639419 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00  
280143 -0.145640 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76  
280149  0.190944  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89  
281144 -0.456108 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00  
281674 -0.072173 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53  

[984 rows x 30 columns]

In [88]:
Y

252624    0
160407    0
155461    0
261270    0
107783    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64

Splitting the dataset into the Training set and Test set

In [89]:
from sklearn.model_selection import train_test_split

In [119]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,stratify=Y,random_state=1)

In [120]:
print("Feature data size: ",X.shape)
print("Training data size: ",X_train.shape)
print("Testing data size: ",X_test.shape)

Feature data size:  (984, 30)
Training data size:  (787, 30)
Testing data size:  (197, 30)


Using Logistic Regression for binary classification.

In [121]:
from sklearn.linear_model import LogisticRegression

TRAINING MODEL WITH TRAINING DATA

In [122]:
from sklearn.metrics import accuracy_score

In [123]:
model=LogisticRegression()

In [124]:
model.fit(X_train,Y_train)

LogisticRegression()

In [125]:
#Finding Accuracy score on training data
X_train_predict=model.predict(X_train)
trainingdata_accuracy_score=accuracy_score(X_train_predict,Y_train)

In [126]:
print('Accuracy score of training data is ',trainingdata_accuracy_score)

Accuracy score of training data is  0.9377382465057179


In [131]:
print('Accuracy percentage of training data: ',trainingdata_accuracy_score*100,'%')

Accuracy percentage of training data:  93.7738246505718 %


In [129]:
#Finding Accuracy score on test data
X_test_predict=model.predict(X_test)
testdata_accuracy_score=accuracy_score(X_test_predict,Y_test)

In [132]:
print('Accuracy score of test data is ',testdata_accuracy_score)

Accuracy score of test data is  0.9289340101522843


In [133]:
print('Accuracy percentage of test data: ',testdata_accuracy_score*100,'%')

Accuracy percentage of test data:  92.89340101522842 %
